controllers/backend/publicdomain/publicdomain.go

```diff
@@ -0,0 +1,120 @@
+		domainLister:     rContext.Global.PublicDomains(settings.RioSystemNamespace).Controller().Lister(),
+		routesetLister:   rContext.Rio.RouteSets("").Controller().Lister(),
+		routesets:        rContext.Rio,
+	}
+	rContext.Global.PublicDomains(settings.RioSystemNamespace).AddLifecycle(ctx, "public-domain-controller", dc)
+}
+
+type domainController struct {
+	namespacesLister corev1.NamespaceLister
+	stackLister      v1beta1.StackLister
+	serviceLister    v1beta1.ServiceLister
+	services         v1beta1.ServicesGetter
+	domains          spacev1beta1.PublicDomainInterface
+	domainLister     spacev1beta1.PublicDomainLister
+	routesetLister   v1beta1.RouteSetLister
+	routesets        v1beta1.RouteSetsGetter
+}
+
+func (d *domainController) Create(domain *spacev1beta1.PublicDomain) (runtime.Object, error) {
+	return domain, nil
+}
+
+func (d *domainController) Updated(domain *spacev1beta1.PublicDomain) (runtime.Object, error) {
+	if err := apply.Apply([]runtime.Object{certs.AcmeIssuer()}, nil, "", "acme-cluster-issuer"); err != nil {
+		return domain, err
+	}
+	ns, err := d.getNamespace(domain)
+	if err != nil {
+		return domain, err
+	}
+	// certificate
+	cert := certs.CertificateHttp(domain)
+	if err := apply.Apply([]runtime.Object{cert}, nil, settings.RioSystemNamespace, "certificate"); err != nil {
+		return domain, err
+	}
+	service, err := d.serviceLister.Get(ns, domain.Spec.TargetName)
+	if err != nil && !errors.IsNotFound(err) {
+		return domain, err
+	}
+	if err == nil {
+		service.Annotations[istio.PublicDomainAnnotation] = domain.Spec.DomainName
+		_, err = d.services.Services(ns).Update(service)
+		return domain, err
+	}
+	routeset, err := d.routesetLister.Get(ns, domain.Spec.TargetName)
+	if err != nil {
+		return domain, err
+	}
+	routeset.Annotations[istio.PublicDomainAnnotation] = domain.Spec.DomainName
+	_, err = d.routesets.RouteSets(ns).Update(routeset)
+	return domain, err
+}
+
+func (d *domainController) Remove(domain *spacev1beta1.PublicDomain) (runtime.Object, error) {
+	ns, err := d.getNamespace(domain)
+	if err != nil {
+		return domain, err
+	}
+	service, err := d.serviceLister.Get(ns, domain.Spec.TargetName)
+	if err != nil && !errors.IsNotFound(err) {
+		return domain, err
+	}
+	if err == nil {
+		delete(service.Annotations, istio.PublicDomainAnnotation)
+		_, err = d.services.Services(ns).Update(service)
+		return domain, err
+	}
+	routeset, err := d.routesetLister.Get(ns, domain.Spec.TargetName)
+	if err != nil {
+		return domain, err
+	}
+	delete(routeset.Annotations, istio.PublicDomainAnnotation)
+	_, err = d.routesets.RouteSets(ns).Update(routeset)
+	return domain, err
+}
+
+func (d *domainController) getNamespace(domain *spacev1beta1.PublicDomain) (string, error) {
+	namespaces, err := d.namespacesLister.List("", labels.Everything())
+	if err != nil {
+		return "", err
+	}
+	for _, ns := range namespaces {
+		if ns != nil && ns.Labels["field.cattle.io/displayName"] == domain.Spec.TargetWorkspaceName {
```

Comment:
You can just add this check to the List call before as a selector.  Then you will just get the object you want directly